### ISSUE TITLE
REPLACE  "upsert" style behavior

### ISSUE THREAD

**Issue by jrots** [RedisLabsModules/RediSearch#150]
Hi, 
In the docs it states that REPLACE behaves as an upsert,
`delete` first `insert/ update` afterwards.

Thought I could use this for updating some fields, without passing all of them each time.

```
127.0.0.1:6379> FT.CREATE myIdx4 STOPWORDS 0 SCHEMA title TEXT title2 TEXT
127.0.0.1:6379> ft.add myIdx4 doc6 1 REPLACE FIELDS title2 7
OK
127.0.0.1:6379> ft.add myIdx4 doc6 1 REPLACE FIELDS title 8
OK
```
```
127.0.0.1:6379> ft.search myIdx4 "@title:8"
1) (integer) 1
2) "doc6"
3) 1) "title"
   2) "8"
   3) "title2"
   4) "7"
```
```
127.0.0.1:6379> ft.search myIdx4 "@title2:7"
1) (integer) 0
```

Just find it a bit weird behavior as it get's saved/ upserted as I would expect in the document data but not in the index. 
Atm I probably need to pass all the fields each time if I want to update a document correctly in the index?

**Comment by jrots** [comment]
Would be handy for project im creating (WIP) - https://github.com/jrots/go-mysql-redisearch  -
to index fields that are scattered amongst several mysql tables in one redisearch index 
(based on https://github.com/siddontang/go-mysql-elasticsearch)

**Comment by jrots** [comment]
just saw that https://github.com/RedisLabsModules/RediSearch/issues/132 is asking for the same functionality.

**Comment by dvirsky** [comment]
Yeah, it's not hard to do, if you want to provide a PR we can guide you on how to do it. It will involve loading the old version of the document, merging with the new one and then indexing as usual. 
BTW very cool project idea. Are you using a MySQL pseudo-slave to do this, or are you polling the database yourself?

I did such a project (with plain old redis) in the past, that emulated a MySQL slave and would just react to events as they would come in. It was very useful.

**Comment by jrots** [comment]
Ok will have a look if I can pull that off -). From tests there does seem to be a lot of index overhead involved when using this constant reindexing approach. 
As a new document is fully added each time to the index document. Will have to revisit my indexing strategy. 

Might be stupid question, but can't documents not be seen as memory pointers that can be changed afterwards in the index? 
I know the index append approach is what's done since beginning of times & makes a lot of sense when you're dealing with disk based indices but given that the nature of redis is in memory.. 
could this not be a possible path?

**Comment by dvirsky** [comment]
The reason we have to only append is that the documents are represented in memory with internal incremental ids that are not exposed to the user. This allows for a huge compression ratio - usually 1 byte is used to encode the document id in the index. 

Actual updating would involve a very heavy process of opening the previous version of the document, creating a diff with the new one, deleting records and rewriting records etc. This would not allow real time updates of the index.

Instead what we do is simply create a new incremental id for the document, reindex it at its new state, and later a garbage collector removes unneeded records from the index. This make deleting a document an O(1) operation, and updating a document just as fast as inserting. 

It's orders of magnitude easier to do load and reindex as in #132.

However, there is an easy alterntive for now: have you seen the [FT.ADDHASH command](http://redisearch.io/Commands/#ftaddhash)?
Updating one field of a document it super easy with it:

```
127.0.0.1:6379> FT.CREATE myIdx4 STOPWORDS 0 SCHEMA title TEXT title2 TEXT
127.0.0.1:6379> ft.add myIdx4 doc6 1 FIELDS title2 7
OK
127.0.0.1:6379> HSET doc6 title 8 <-- instead of directly doing FT.ADD
127.0.0.1:6379> ft.addhash myIdx4 doc6 1 REPLACE <---- now we reindex eeverything
OK
```
ADDHASH will read the current state of the document, so if you do the partial updates and then the reindexing, it's exactly what you need. And using a transaction you can do this atomically as well.

**Comment by jrots** [comment]
@dvirsky  
Alright, got it -  that `ft.addhash` makes sense now. 
Thx for the thorough explanation! 
---- 
Other than that what would be in your opinion the best approach in solving the following use case: 
Index with user data with approx. +/- 150M docs and 20+ fields per doc. 
Data can be searched with a lot of range queries, country, gender/ age/ etc filters, most of the data indexed is actually "stale"

We want however, limit/ sort the results by ordering on the "newest", lastonline users first. 
So that lastonline field is very dynamic - as users can come online a lot during the day. 
Even if we limit updates for that field to f.e. every 15 minutes it could quickly generate 20 updates for a user in a day on the index if that user is active for 5 hours during that day. 
So even with only 1M users active, it could result in 20M "hard" reindexes, most likely a lot more.

Haven't tried the full indexing and practical effects yet, but think it will be very heavy for GC + index size.  Will update on numbers once I have fully indexed everything.

Other approaches that would be less "hard" I could think of ..
- have a custom sorter that reads the lastonline field from redis 
(would be less intrusive for the index)
- store the users that were last online in that day for a certain filter in a separte key <gender>_<country>, and use
	INKEYS on the index
- store the results of the search & join them with a zset that keeps track of the lastonline of these users.. 
- create separate indexes based on "day" that a user was online ..

**Comment by dvirsky** [comment]
Hmm.... 
Well, 20M reindexes on small objects is not hard to do, but it's wasteful - and you'll exhaust the 4B available ids in a few months. So you're right that we should try to avoid it if that is the only change. 

There are 3 options I can think of that will let us bypass reindexing, and will require little development on my side:

1. **Using document payloads and a custom scorer**. While currently you can set payloads per document and evaluate them with the scorer (so you can just encode the timestamp as a payload), it is not possible to update the payloads. However that is only an API limitation, internally it doesn't require reindexing. So adding a command like `FT.SETPAYLOAD {idx} {docId} {newPayload}` would solve your issue - albeit require that you write a custom scorer.

2. **Using non indexable sortable fields**. Right now, a sortable field has to be indexed. If I'll add a special sortable field that is not indexed, and a special update command - you could simply update this field only and use SORTBY with no need to reindex. This is very simple to do internally (indexing and sorting by fields are two separate mechanisms) but feels like it's creating a clunky API. But it's certainly doable and not hard.

3. **Using the document score**. You can encode the timestamp in the document score (even though it has to be between 0 and 1, it's double precision and easy to encode the timestamp as a fraction - simply add a decimal point ahead of it). and like the payload, we can add a simple command (it's actually in the TODO since it's very useful for other stuff) like `FT.SETSCORE {idx} {docId} {newScore}` that will update it without reindexing. You would have to write a special scorer for this because the actual scorer uses the document score factored in with other stuff. But since other people might want this, adding a built in scorer that just returns the document score is easy. 

So I would say 3 is the most compelling and easiest for us both. And while I'm doing 3 I'll probably add 1 as well since it's basically the same logic, so it will be your choice which to use. 

WDYT?

**Comment by dvirsky** [comment]
Another cool option since sorting by time is a common need - is to add a timestamp as a built in property of each document, just like the score.

**Comment by jrots** [comment]
All look like valid solutions imho.
From a practical perspective I like 2) the most .. might make the api not as lean, 
but just need to find a good name/ api entry point for it.  
It gives a lot of versatility/ flexibility back to the users.. 
If I for example would want to have a "sortable" field 
that is composed out of different stuff (fe. the ones with an avatar first + afterwards lastonline),  
I could easily generate/ regenerate a new index without writing custom C functions (that potentially leak and are slower). 
I also have a field called "hotness" that is calculated each day on the warehouse side of things, 
& would require a reindex to update that and to sort on that. I don't have to write a custom scorer in C & compile/ restart redis as it would be possible with that api. 
.. 
you could also expose it as `ADD SORTINDEX idxName( 'hasavatar', 'lastonline')` f.e. and have all the internal updating happen in the back.. 
Though having a command as `FT.UPDATE_SORTBY "somefield" 1210011` is equally good for me  

But then again, pick the one you think would be the  best because I'm just speaking out for this/my use case :).

**Comment by dvirsky** [comment]
I'm trying to be guided by users. I suppose the UPDATE command could be smart and detect a situation where nothing that is updated is an indexed property, and then do a light update and not a full reindex.

**Comment by jrots** [comment]
Yes that would be the best - some flag that a field is only for sorting.

**Comment by dvirsky** [comment]
I suppose a single update command could cover all 3 options actually. 

FT.UPDATE index docid [SCORE 1] [PAYLOAD foo] [FIELDS ...]

So for score and payload we just update them, and for field we check them to see if at least one needs reindexing. If that is the case we do a full reindex. otherwise we just update the hash and the sorting vector.

**Comment by dvirsky** [comment]
We can go even further and add a keyword like PARTIAL to FT.ADD.

**Comment by jrots** [comment]
Cool, looks like clean solution! 
What would happen if you pass "real" indexed fields to `FT.ADD PARTIAL` probably just do a real `ADD` to the index for those fields?

**Comment by dvirsky** [comment]
In the case of REPLACE only - we behave like today.

In the case of REPLACE + PARTIAL we do the following:
1. do HMSET on the values
2. If any of the provided fields are indexed:
   a. load the old version of the document and merge with the partial one
   b. reindex the entire document
2. If a score is set - update that (regardless of whether we indexed or not)
3. If a payload is set - update that (regardless of whether we indexed or not)
4. If any sortable fields were provided - udpate the sorting vector, regardless of whether we indexed or not.
